src/oauth2.rs

```diff
@@ -1,576 +1,416 @@
-            match serde_json::to_string(&token_info) {
-                Ok(token_info_string) => {
-                    self.save_token_info(&token_info_string);
-                    Some(token_info)
-                }
-                Err(why) => {
-                    panic!(
-                        "couldn't convert token_info to string: {} ",
-                        why.to_string()
-                    );
-                }
-            }
-        } else {
-            None
-        }
+    /// The same as `request_client_token_without_cache`, but saves the token
+    /// into the cache file if possible.
+    #[maybe_async]
+    pub async fn request_client_token(&mut self) -> ClientResult<()> {
+        self.request_client_token_without_cache().await?;
+        self.write_token_cache()
     }
-    fn save_token_info(&self, token_info: &str) {
-        save_token_info(token_info, self.cache_path.as_path())
+
+    /// Parse the response code in the given response url. If the URL cannot be
+    /// parsed or the `code` parameter is not present, this will return `None`.
+    ///
+    /// Step 2 of the [Authorization Code Flow
+    /// ](https://developer.spotify.com/documentation/general/guides/authorization-guide/#authorization-code-flow).
+    pub fn parse_response_code(&self, url: &str) -> Option<String> {
+        let url = Url::parse(url).ok()?;
+        let mut params = url.query_pairs();
+        let (_, url) = params.find(|(key, _)| key == "code")?;
+        Some(url.to_string())
     }
-    fn is_scope_subset(needle_scope: &mut str, haystack_scope: &mut str) -> bool {
-        let needle_vec: Vec<&str> = needle_scope.split_whitespace().collect();
-        let haystack_vec: Vec<&str> = haystack_scope.split_whitespace().collect();
-        let needle_set: HashSet<&str> = HashSet::from_iter(needle_vec);
-        let haystack_set: HashSet<&str> = HashSet::from_iter(haystack_vec);
-        // needle_set - haystack_set
-        needle_set.is_subset(&haystack_set)
+
+    /// Obtains the user access token for the app with the given code without
+    /// saving it into the cache file, as part of the OAuth authentication.
+    /// The access token will be saved inside the Spotify instance.
+    ///
+    /// Step 3 of the [Authorization Code Flow](https://developer.spotify.com/documentation/general/guides/authorization-guide/#authorization-code-flow).
+    #[maybe_async]
+    pub async fn request_user_token_without_cache(&mut self, code: &str) -> ClientResult<()> {
+        let oauth = self.get_oauth()?;
+        let mut data = Form::new();
+        data.insert("grant_type".to_owned(), "authorization_code".to_owned());
+        data.insert("redirect_uri".to_owned(), oauth.redirect_uri.clone());
+        data.insert("code".to_owned(), code.to_owned());
+        data.insert("scope".to_owned(), oauth.scope.clone());
+        data.insert("state".to_owned(), oauth.state.clone());
```

Comment:
Perhaps we should replace these hardcoded string `grant_type`, `redirect_uri `, `code` with constant variable or enum either.

Reply:
done